### PR TITLE
utils.memory: create a meminfo object

### PR DESCRIPTION
**Pull request by apahim**
This module has a number of functions aiming to retrieve and/or
convert values from /proc/meminfo. Feels like it's time to create
a more robust representation of /proc/meminfo. This commit creates
a MemInfo class, instantiated in a meminfo object.

The MemInfo class carries one attribute for each item in /proc/meminfo.
Each attribue is in fact an instance of a MemInfoObject class, which
contains attribbutes to expose the item value. The basic 'value'
attribute will show the value as is. The additional 'value_in_*'
attributes will show the converted value, according to the selected
unit.

Usage:

    >>> from avocado.utils import memory

    >>> memory.meminfo
    <avocado.utils.memory.MemInfo object at 0x7fe8e737f250>

    >>> dir(memory.meminfo)
    ['Active', 'Active_anon_', 'Active_file_', 'AnonHugePages',
    'AnonPages', 'Bounce', 'Buffers', 'Cached', 'CmaFree', 'CmaTotal',
    'CommitLimit', 'Committed_AS', 'DirectMap1G', 'DirectMap2M',
    'DirectMap4k', 'Dirty', 'HardwareCorrupted', 'HugePages_Free',
    'HugePages_Rsvd', 'HugePages_Surp', 'HugePages_Total', 'Hugepagesize',
    'Inactive', 'Inactive_anon_', 'Inactive_file_', 'KernelStack',
    'Mapped', 'MemAvailable', 'MemFree', 'MemInfoObject', 'MemTotal',
    'Mlocked', 'NFS_Unstable', 'PageTables', 'SReclaimable', 'SUnreclaim',
    'Shmem', 'ShmemHugePages', 'ShmemPmdMapped', 'Slab', 'SwapCached',
    'SwapFree', 'SwapTotal', 'Unevictable', 'VmallocChunk', 'VmallocTotal',
    'VmallocUsed', 'Writeback', 'WritebackTmp', '_MemInfo__meminfo',
    '__class__', '__delattr__', '__dict__', '__dir__', '__doc__',
    '__format__', '__getattr__', '__getattribute__', '__hash__',
    '__init__', '__iter__', '__module__', '__new__', '__reduce__',
    '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__',
    '__subclasshook__', '__weakref__']

    >>> memory.meminfo.MemTotal
    <avocado.utils.memory.MemInfoObject object at 0x7fe8f6ff8750>

    >>> dir(memory.meminfo.MemTotal)
    ['_MemInfoObject__units', '__call__', '__class__', '__delattr__',
    '__dict__', '__dir__', '__doc__', '__format__', '__getattr__',
    '__getattribute__', '__hash__', '__init__', '__module__', '__new__',
    '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__',
    '__str__', '__subclasshook__', '__weakref__', '_convert', 'name',
    'value', 'value_in_bytes', 'value_in_gbytes', 'value_in_kbytes',
    'value_in_mbytes', 'value_in_tbytes']

    >>> memory.meminfo.MemTotal.value
    15853856

    >>> memory.meminfo.MemTotal.value_in_mbytes
    15482

    >>> for item in memory.meminfo:
    ...     print item.name, item.value_in_mbytes
    ...
    WritebackTmp 0
    SwapTotal 7744
    SwapFree 7744
    DirectMap4k 384
    KernelStack 5
    MemFree 10785
    HugePages_Rsvd 0
    Committed_AS 2108
    SUnreclaim 197
    NFS_Unstable 0
    VmallocChunk 0
    CmaFree 0
    Writeback 0
    MemTotal 15482
    VmallocUsed 0
    DirectMap1G 6144
    HugePages_Free 0
    AnonHugePages 0
    Active_file_ 1259
    AnonPages 440
    Inactive_anon_ 67
    Active 1719
    CommitLimit 15486
    Hugepagesize 2
    Active_anon_ 459
    Cached 2969
    SwapCached 0
    VmallocTotal 33554431
    CmaTotal 0
    Shmem 87
    Mapped 242
    ShmemPmdMapped 0
    Unevictable 0
    SReclaimable 536
    MemAvailable 13874
    Mlocked 0
    DirectMap2M 10310
    HugePages_Surp 0
    Bounce 0
    Inactive 1691
    PageTables 19
    HardwareCorrupted 0
    ShmemHugePages 0
    HugePages_Total 0
    Slab 734
    Inactive_file_ 1623
    Buffers 2
    Dirty 0

Signed-off-by: Amador Pahim <apahim@redhat.com>

### CODE REVIEW

**Comment by apahim** [comment]
@abdhaleegit, @PraveenPenguin, @narasimhan-v, what about this PR? Do you like it?

File: avocado/utils/memory.py [hunk 1]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+        Representation of one item from /proc/meminfo
+        """
+        def __init__(self, name, value):
+            self.name = name
+            self.value = int(value)
+            self.__units = {'bytes': 2**0,
+                            'kbytes': 2**10,
+                            'mbytes': 2**20,
+                            'gbytes': 2**30,
+                            'tbytes': 2**40}
```
Comment (ldoktor) [inline avocado/utils/memory.py]: Well this is nice, but it requires 5x power to be evaluated on each MemInfoObject creation despite whether it'd ever beconverted or not. I'd prefer good old boring hardcoded methods which does not require any workarounds in __getattr_, no units dicts etc.

File: avocado/utils/memory.py [hunk 2]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+
+        def __call__(self):
+            """
+            Returns the value when the class is called.
+            """
+            return self.value
```
Comment (ldoktor) [inline avocado/utils/memory.py]: I don't like the idea of calling an object to get value in kb. Not sure how others feel about this...

File: avocado/utils/memory.py (hunk 1 above)
Comment (ldoktor) [inline avocado/utils/memory.py]: `MemInfoObject.bytes`, `MemInfoObject.kilobytes`, `MemInfoObject.megabytes`, `MemInfoObject.gigabytes` properties, perhaps with shortcuts to `b`, `kb`, `mb`, `gb`, therefor in test one would use:

```python
memory.meminfo.MemFree.kb
```

Comment (ldoktor) [inline avocado/utils/memory.py]: .. or just the `b`, `kb`, ...

File: avocado/utils/memory.py [hunk 3]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+                raise TypeError('Unit %s is not valid.')
+            return int(mem) * self.__units['kbytes'] / self.__units[unit]
+
+    def __init__(self):
+        self.__meminfo = {}
+        with open('/proc/meminfo', 'r') as f:
+            for line in f.readlines():
+                item, value = line.strip().split(None, 1)
+                item = item.replace('(', '_').replace(')', '_').strip(':')
+                value = value.strip(' kB')
```
Comment (ldoktor) [inline avocado/utils/memory.py]: strip is dangerous, it removes all ` `, `k` and `B`s. Not just the tailing ` kB`. Not even `rstrip` does that. So instead I'd go with `[:-3]`, or with a regexp. A compiled regexp might be (not tested) actually faster then all the plain striping and splitting.

Comment (ldoktor) [inline avocado/utils/memory.py]: Alternatively good old `split()` should do as first item is always key, second value and third units.

File: avocado/utils/memory.py [hunk 4]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+                self.__meminfo[item] = value
+
+    def __getattr__(self, name):
+        """
+        Creates one extra attribute per item in /proc/meminfo.
+        """
+        try:
+            return self.MemInfoObject(name, self.__meminfo[name])
+        except KeyError:
```
Comment (ldoktor) [inline avocado/utils/memory.py]: As you are not really doing any exception-handling, just masking the KeyError with AttributeError the use of `if name in self.__meminfo` would IMO be more appropriate.

File: avocado/utils/memory.py [hunk 5]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+        listing = dir(type(self)) + list(self.__dict__.keys())
+        listing.extend([item for item in self.__meminfo])
+        return listing
+
+    def __iter__(self):
+        for item in self.__meminfo:
+            yield self.MemInfoObject(item, self.__meminfo[item])
+
+
+meminfo = MemInfo()
```
Comment (ldoktor) [inline avocado/utils/memory.py]: This means first `import memory` will attempt to read `/proc/meminfo` even when we do not actually need it. Not really happy about it...

File: avocado/utils/memory.py [hunk 6]
```diff
@@ -320,3 +320,87 @@ def get_thp_value(feature):
...
+                item = item.replace('(', '_').replace(')', '_').strip(':')
+                value = value.strip(' kB')
+                self.__meminfo[item] = value
+
+    def __getattr__(self, name):
+        """
+        Creates one extra attribute per item in /proc/meminfo.
+        """
+        try:
+            return self.MemInfoObject(name, self.__meminfo[name])
```
Comment (ldoktor) [inline avocado/utils/memory.py]: Well this is actually very wrong. Wast majority of the attributes are dynamic and people are interested in the current value, not in the value which was there on `import` time.

**Comment by ldoktor** [comment]
Btw the simplest solution would be to simply generate:

```python
class __Item(object):
    def __init__(self, key):
        self.key = key
    def value(self):
        return read_from_meminfo(self.key)
    @property
    def b(self):
        return self.__value() * 1024
    @property
    def kb(self):
        return self.__value()
    @property
    def mb(self):
        return self.__value() / 1024
...

class MemInfo(object):
    def __init__(self):
        for line in open("/proc/meminfo"):
            key = line.split(None, 1)[0].replace...
            setattr(self, key, __Item(key))

meminfo = MemInfo()
```

And ideally to lazily set the attrs on first `getattr`...

**Pull request closed without merging**